### ISSUE TITLE
[Placeholder] Fine grained RBAC

### ISSUE THREAD

**Issue by e0d** [openedx/platform-roadmap#246]

**Comment by jmakowski1123** [comment]
Bryan Kerston (2U) is volunteering to lead this initiative

**Comment by bryan-kersten** [comment]
My team is looking to lead work on this project.

**Comment by itsjeyd** [comment]
> Paulo and ASU and Bryan coordinating on subset of features required for ASU

Yep! OpenCraft is planning to follow 2U's activities around this project, and to keep @bryan-kersten and his team in the loop on relevant PRs for ASU. We'll be looking for feedback on any aspects that might clash with 2U's plans and/or could be implemented in a different way that would better support future efforts and refactorings from Bryan's team.

**Comment by bryan-kersten** [comment]
Hi @itsjeyd Can you share here the specific PRs that you'll be completing for ASU? Just want to make sure we have that noted as we work towards larger project.

**Comment by itsjeyd** [comment]
Hey @bryan-kersten, thanks for checking in! The PRs that OpenCraft submitted for ASU so far are:

* https://github.com/openedx/edx-platform/pull/32015
* https://github.com/openedx/edx-platform/pull/32097

_@0x29a If the list above is missing something, please let me know._

In terms of additional features related to fine-grained RBAC, we're back to the drawing board for now: ASU found a way to leverage existing functionality to address some of their needs in the short term. So the next step will be to iterate on their requirements for new capabilities and define more precisely what they would like us to add to the platform.

We'll keep you in the loop on that.

In the meantime, if you have any design docs or specs for the functionality that your team is planning to implement, please let us know.

CC @cassiezamparini @viadanna

**Comment by bryan-kersten** [comment]
Thanks Tim. Really appreciate you sharing this info and keeping me in loop as requirements iterate. We are in process of gathering a larger understanding of the problem set around RABC. As we get that together I will share with you here so we can be sure to capture the broader sentiment of the community before we hone in on prioritization and approach.

**Comment by itsjeyd** [comment]
That sounds great @bryan-kersten, thanks for the update!

CC @cassiezamparini @viadanna @0x29a

**Comment by jmakowski1123** [comment]
Bryan's RBAC Problem Statement: https://docs.google.com/document/d/1Imgf3wXpJaoc_dw-jJwBLdkRWHNLlHyTScaNJoJQOic/edit#

**Comment by itsjeyd** [comment]
For latest batch of updates about this epic see [7-21-2023 - Campus Working Group Meeting Notes](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3822354433/7-21-2023+-+Campus+Working+Group+Meeting+Notes) (first item).

**Comment by hsinkoff** [comment]
[Tech Spec](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3824648277/RBAC+Tech+Spec) related to the RBAC project.

**Comment by hsinkoff** [comment]
Update:The Tech Spec has been posted to [Slack](https://app.slack.com/client/T02SNA1T6/C12M8M5AR/thread/C12M8M5AR-1690555081.146809) and [Discourse](https://discuss.openedx.org/t/tech-spec-roles-and-permissions-rbac/10830). We have not received any feedback since posting it, but have requested all feedback be provided by August 23rd, 2023. 

At this time the plan is to move forward with adding a new course roles system beginning after the 23rd. We will incorporate any feedback received before proceeding and then will move forward with adding the new system, first for usage with a new role and if successful we will begin switching existing roles to the new system.

In addition to the technical plan, we have created a permission grid with a list of around 20 permissions that would be assignable to a role. The list allows for the ability to recreate all existing `student_courseaccessrole` roles and all `django_comment_client_role` roles with no planned changes to their current usage.

**Comment by hsinkoff** [comment]
Update:
The Tech Spec has been posted to [Slack](https://app.slack.com/client/T02SNA1T6/C12M8M5AR/thread/C12M8M5AR-1690555081.146809) and [Discourse](https://discuss.openedx.org/t/tech-spec-roles-and-permissions-rbac/10830). We have not received any feedback since posting it, but have requested all feedback be provided by August 23rd, 2023.

At this time the plan is to move forward with adding a new course roles system beginning after the 23rd. We will incorporate any feedback received before proceeding and then will move forward with adding the new system; first for usage with a new role(s) and then, if this option proves successful, we will begin switching existing roles to the new system.

In addition to the technical plan, we have created a [permission grid](https://docs.google.com/spreadsheets/d/1htsV0eWq5-y96DZ5A245ukfZ4_qeH0KjHVaOyfqD8OA/edit#gid=0) with a list of around 20 permissions that would be assignable to a role. Please review these permissions and provide any feedback you may have to bbrown1@2u.com by Friday September 1st, 2023.

Some things of note while reviewing the proposed permissions:
- When possible permissions should span features, feature-specific permissions will be used only when absolutely necessary
- [Permissions](https://docs.google.com/spreadsheets/d/1htsV0eWq5-y96DZ5A245ukfZ4_qeH0KjHVaOyfqD8OA/edit#gid=0) are stacked to [create roles](https://docs.google.com/spreadsheets/d/1htsV0eWq5-y96DZ5A245ukfZ4_qeH0KjHVaOyfqD8OA/edit#gid=908503896)
- Tab 2 ([r&p](https://docs.google.com/spreadsheets/d/1htsV0eWq5-y96DZ5A245ukfZ4_qeH0KjHVaOyfqD8OA/edit#gid=908503896)) shows the mapping for all current student_courseaccessrole and django_comment_client role roles and three potential new roles

**Comment by mariajgrimaldi** [comment]
Thanks, @hsinkoff, for sharing the proposal with @felipemontoya and me I hope this is still useful, thanks for the patience!

We came up with an idea after carefully studying the requirements of the Spanish Consortium Project (the initiative behind the [Flexible Groups requirement](https://github.com/openedx/platform-roadmap/issues/250)) alongside the new systems' proposal. Since the flexible group requirement strongly inclines towards having an instructor-like course role that manages a group of students, we came up with a new set of permissions for you folks to consider:  

A set of permissions (like `Manage students` and `Manage gradebook`) but the students belong to a cohort (or any other grouping strategy). Here are some examples of what a user with this permission could do:

The user with this permission who belongs to cohort A could:
- Grade a student if and only if the student belongs to that cohort.
- Reset students' attempts if and only if the student belongs to that cohort 
And so on.

This permission denies access to resources that don't belong to that grouping. Or even, instead of just considering grouping strategies, this permission could behave like a `Manage students if <cond>` where _cond_ could be:

- If belongs to a cohort
- If speaks English
- If belongs to a team

And so on. This way, we can filter the dataset the user can actually interact with. We think this could be easily implemented by adding Open edX Filters that _filters_ the students' dataset after checking for permissions.

Please, let us know what you think. We're willing to collaborate to move this forward if necessary. Thanks!

**Comment by hsinkoff** [comment]
@mariajgrimaldi Thank you for reviewing the proposal. 

We are moving forward with the MVP and Phase 1 of the project at this time. 

It was decided that the MVP will include permissions that replicate the existing access levels that roles provide and that Phase 1 would include the work needed to transition existing roles into the new system. The system is being designed in a way that will give it flexibility for adding new permissions that correspond to new functionality at a later date. 

The permission Maria and Felipe have proposed sounds like new functionality so it will not fit into the guidelines for our MVP or Phase 1, but we do think it sounds like a good candidate for a future permission once the MVP and Phase 1 are complete. 

Happy to follow-up after we’ve delivered the new permissions system.

**Comment by felipemontoya** [comment]
@hsinkoff thanks a lot for reviewing @mariajgrimaldi's proposal. We would be very interested in pursuing this, would you be open to a collaboration where we write the code for the new role and the necessary functionality to make it land during the phase 1? or shortly after if part of the MVP is definitely not possible.

**Comment by hsinkoff** [comment]
@felipemontoya, yes we would be open to collaborating with you on this. I'll reach out to you in Slack to set up some time to discuss the specific details and make sure we're all on the same page.

**Comment by hsinkoff** [comment]
**Status Update**

Implementation work is in progress with the current focus on model setup and adding permission checking to all current locations where roles are checked. Work is also being done to finalize [names and descriptions for the permissions](https://docs.google.com/spreadsheets/d/1htsV0eWq5-y96DZ5A245ukfZ4_qeH0KjHVaOyfqD8OA/edit#gid=0) that will be used to build the existing (and new) roles.

The initial work will be additive and will check permissions in addition to the existing roles. Work is being completed on a [feature branch](https://github.com/openedx/edx-platform/tree/ROLES-2-course_roles_setup).

**Comment by hsinkoff** [comment]
**Status Update**

The [CourseRoles service PR](https://github.com/openedx/edx-platform/pull/33734) is up for review. This will be the first of many PRs required for this work, but will set up the foundational requirements of a new roles and permissions system for course level roles. Within the next month we anticipate additional PRs (in both edx-platform and frontend-app-course-authoring) that add functionality and permissions usage.

**Comment by hsinkoff** [comment]
**Status Update**

Work on this project is ongoing. There are no new changes to report on the status of the work.

**Comment by jristau1984** [comment]
Hello Open edX Community -

The 2U team currently working on the CourseRoles project is pausing their work for the foreseeable future. This pause is happening because additional role granularity is no longer prioritized. If any community members are interested and able to pick up this project, please consider working on it.

All work related to the project is in two feature branches, one in the [edx-platform](https://github.com/openedx/edx-platform/tree/CourseRoles) repo and one in the [frontend-app-course-authoring](https://github.com/openedx/frontend-app-course-authoring/tree/CourseRoles) repo. 

Documentation relating to the current technical progress of the project and open questions has been added to the [docs folder](https://github.com/openedx/edx-platform/tree/CourseRoles/openedx/core/djangoapps/course_roles/docs) in the edx-platform branch. Additionally documentation about checking access has been added in the frontend-app-course-authoring [docs/authorization folder](https://github.com/openedx/frontend-app-course-authoring/tree/CourseRoles/docs/authorization/permissions_check_architecture). 

Axim has more detailed next step options for consideration if an individual or team chooses to pick up this project. The next steps should be chosen in consultation with Axim and the campus working group.

**Comment by jmakowski1123** [comment]
We are moving forward with an MVP for RBAC in Ulmo. Documentation here: https://openedx.atlassian.net/wiki/spaces/OEPM/pages/5032345665/RBAC+MVP+Libraries

**Comment by robrap** [comment]
Thanks @jmakowski1123. Consider whether the new system will replace an older mechanism in any way that the DEPR process might be needed to inform and manage. Thanks.
